### PR TITLE
Modify the core to support removing the necessity of having role_org-application association

### PR DESCRIPTION
**Pull request by DMHP**
### Proposed changes in this pull request
Modify the core to support removing the necessity of having role_org-application association

### When should this PR be merged
Immediately

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ApplicationBasicInfo.java [hunk 1]
```diff
@@ -238,4 +239,21 @@ public void setTenantDomain(String tenantDomain) {
 
         this.tenantDomain = tenantDomain;
     }
+
+    /**
+     * Get application uuid.
+     * @return uuid application uuid
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ApplicationBasicInfo.java:245]:
```suggestion
     * Get application uuid.
     *
     * @return uuid application uuid.
```

Comment (AnuradhaSK) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ApplicationBasicInfo.java:245]: Check formatting in other places as well

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java [hunk 2]
```diff
@@ -116,6 +116,22 @@ public abstract ApplicationBasicInfo[] getAllApplicationBasicInfo(String tenantD
     public abstract ApplicationBasicInfo[] getApplicationBasicInfo(String tenantDomain, String username, String filter)
             throws IdentityApplicationManagementException;
 
+    /**
+     *
```
Comment (shashimalcse) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java:120]: lets add the method description

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 3]
```diff
@@ -2243,22 +2252,39 @@ private AssociatedRolesConfig getAssociatedRoles(String applicationId, Connectio
...
+                throw new IdentityApplicationManagementException("Error while retrieving associated roles for " +
+                        "application ID: " + applicationId, e);
+            }
+
         }
-        String allowedAudience =
-                getSPPropertyValueByPropertyKey(applicationId, ALLOWED_ROLE_AUDIENCE_PROPERTY_NAME, tenantDomain);
+
         associatedRolesConfig.setAllowedAudience(
                 StringUtils.isNotBlank(allowedAudience) ? allowedAudience.toLowerCase() : RoleConstants.ORGANIZATION);
```
Comment (shashimalcse) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:2289]: we set the default audience as organization for the application with these changes?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java [hunk 4]
```diff
@@ -78,6 +78,15 @@ public class ApplicationMgtDBQueries {
             "TENANT_ID = ? AND APP_NAME != ? ORDER BY ID DESC";
     public static final String LOAD_APP_NAMES_BY_TENANT_AND_APP_NAME = "SELECT ID, APP_NAME, DESCRIPTION FROM SP_APP " +
             "WHERE TENANT_ID = ? AND APP_NAME != ? AND (%s) ORDER BY ID DESC";
+
+    public static final String LOAD_APP_IDS_BY_SP_PROPERTY_H2 = "SELECT SP_APP.UUID FROM SP_APP  JOIN " +
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java:82]:
```suggestion
    public static final String LOAD_APP_IDS_BY_SP_PROPERTY_H2 = "SELECT SP_APP.UUID FROM SP_APP JOIN " +
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java [hunk 5]
```diff
@@ -78,6 +78,15 @@ public class ApplicationMgtDBQueries {
...
+            "SP_METADATA  ON SP_APP.ID = SP_METADATA.SP_ID WHERE SP_METADATA.NAME=? and " +
+            "SP_METADATA.`VALUE`=? AND SP_METADATA.TENANT_ID = ?";
+
+    public static final String LOAD_APP_IDS_BY_SP_PROPERTY = "SELECT SP_APP.UUID FROM SP_APP  JOIN " +
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java:86]:
```suggestion
    public static final String LOAD_APP_IDS_BY_SP_PROPERTY = "SELECT SP_APP.UUID FROM SP_APP JOIN " +
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java [hunk 6]
```diff
@@ -199,6 +199,20 @@ default ApplicationBasicInfo getApplicationBasicInfoByResourceId(String resource
         return null;
     }
 
+    /**
+     * Retrieve application basic information using the sp metadata property key and value.
+     *
+     * @param key Name of the sp metadata property key
+     * @param value Value of the sp metadata property
+     * @return ApplicationBasicInfo containing the basic app information
+     * @throws IdentityApplicationManagementException
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java:208]: complete the exception description

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 7]
```diff
@@ -193,6 +195,8 @@ public class ApplicationDAOImpl extends AbstractApplicationDAOImpl implements Pa
 
     private static final String SP_PROPERTY_NAME_CERTIFICATE = "CERTIFICATE";
     private static final String APPLICATION_NAME_CONSTRAINT = "APPLICATION_NAME_CONSTRAINT";
+    public static final String AUDIENCE_EQ_ORGANIZATION = "audience eq organization";
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:198]: Shall we build this filter value by combining existing constants for audience, eq and organization separately

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 8]
```diff
@@ -2243,22 +2252,39 @@ private AssociatedRolesConfig getAssociatedRoles(String applicationId, Connectio
...
-            throw new IdentityApplicationManagementException(
-                    "Error while retrieving associated roles for application ID: " + applicationId, e);
+        } else if (RoleConstants.ORGANIZATION.equalsIgnoreCase(allowedAudience)) {
+            RoleManagementService roleManagementServiceV2 = ApplicationManagementServiceComponentHolder.getInstance()
+                    .getRoleManagementServiceV2();
+            try {
+                List<RoleBasicInfo> listOfAllRoles = roleManagementServiceV2.getRoles(AUDIENCE_EQ_ORGANIZATION,
+                        0, 0, null, null, tenantDomain);
+                List<String> roleIds = listOfAllRoles.stream().map(RoleBasicInfo::getId).collect(Collectors.toList());
+                associatedRolesConfig.setRoles(buildAssociatedRolesWithRoleName(roleIds, tenantDomain));
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:2280]: Are we returning all organization audienced roles as the associated roles, if the app is using organization audienced roles?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 9]
```diff
@@ -3747,6 +3773,54 @@ public int getCountOfApplications(String filter) throws IdentityApplicationManag
         return count;
     }
 
+    /**
+     * Retrieve application basic information using the sp metadata property key and value.
+     *
+     * @param key Name of the sp metadata property key
+     * @param value Value of the sp metadata property value
+     * @return
+     * @throws IdentityApplicationManagementException
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3782]: complete

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 10]
```diff
@@ -3747,6 +3773,54 @@ public int getCountOfApplications(String filter) throws IdentityApplicationManag
...
+     */
+    @Override
+    public ApplicationBasicInfo[] getApplicationBasicInfoBySPProperty(String key, String value)
+            throws IdentityApplicationManagementException {
+
+        int tenantID = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+        if (log.isDebugEnabled()) {
+            log.debug("Getting the all applications for the tenant: " + tenantID);
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3790]:
```suggestion
            log.debug(String.format("Getting the all applications matching to property: %s value equals to %s in the tenant:  %d", key, value, tenantID);
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 11]
```diff
@@ -3747,6 +3773,54 @@ public int getCountOfApplications(String filter) throws IdentityApplicationManag
...
+        }
+        Connection connection = IdentityDatabaseUtil.getDBConnection(false);
+        PreparedStatement getAppNamesStmt = null;
+        ResultSet appNameResultSet = null;
+        ArrayList<ApplicationBasicInfo> appInfo = new ArrayList<ApplicationBasicInfo>();
+
+        try {
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3797]: Shall we use try with resource, making connection and preparedStatement resources which auto closable at the end of usage?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java [hunk 12]
```diff
@@ -78,6 +78,15 @@ public class ApplicationMgtDBQueries {
             "TENANT_ID = ? AND APP_NAME != ? ORDER BY ID DESC";
     public static final String LOAD_APP_NAMES_BY_TENANT_AND_APP_NAME = "SELECT ID, APP_NAME, DESCRIPTION FROM SP_APP " +
             "WHERE TENANT_ID = ? AND APP_NAME != ? AND (%s) ORDER BY ID DESC";
+
+    public static final String LOAD_APP_IDS_BY_SP_PROPERTY_H2 = "SELECT SP_APP.UUID FROM SP_APP  JOIN " +
+            "SP_METADATA  ON SP_APP.ID = SP_METADATA.SP_ID WHERE SP_METADATA.NAME=? and " +
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java:83]:
```suggestion
            "SP_METADATA ON SP_APP.ID = SP_METADATA.SP_ID WHERE SP_METADATA.NAME=? and " +
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java [hunk 13]
```diff
@@ -78,6 +78,15 @@ public class ApplicationMgtDBQueries {
...
+            "SP_METADATA.`VALUE`=? AND SP_METADATA.TENANT_ID = ?";
+
+    public static final String LOAD_APP_IDS_BY_SP_PROPERTY = "SELECT SP_APP.UUID FROM SP_APP  JOIN " +
+            "SP_METADATA  ON SP_APP.ID = SP_METADATA.SP_ID WHERE SP_METADATA.NAME=? and " +
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java:87]:
```suggestion
            "SP_METADATA ON SP_APP.ID = SP_METADATA.SP_ID WHERE SP_METADATA.NAME=? and " +
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java (hunk 8 above)
Comment (DMHP) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:2280]: Yes

**Review by shashimalcse — DISMISSED** [review]

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 14]
```diff
@@ -2243,22 +2251,45 @@ private AssociatedRolesConfig getAssociatedRoles(String applicationId, Connectio
...
-        } catch (SQLException | IdentityRoleManagementException e) {
-            throw new IdentityApplicationManagementException(
-                    "Error while retrieving associated roles for application ID: " + applicationId, e);
+        } else if (RoleConstants.ORGANIZATION.equalsIgnoreCase(allowedAudience)) {
+            ApplicationManagementServiceComponentHolder holder = ApplicationManagementServiceComponentHolder.
+                    getInstance();
+            RoleManagementService roleManagementService = holder.getRoleManagementServiceV2();
+            try {
+                if (roleManagementService != null) {
+                    List<RoleBasicInfo> listOfAllRoles = roleManagementService.
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:2278]:
```suggestion
                    List<RoleBasicInfo> listOfAllOrgRoles = roleManagementService.
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 15]
```diff
@@ -2243,22 +2251,45 @@ private AssociatedRolesConfig getAssociatedRoles(String applicationId, Connectio
...
+                            getRoles(RoleConstants.AUDIENCE + SPACE + RoleConstants.EQ + SPACE +
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:2279]: Is there any max limit defined in role mgt service when request to get all roles matching to a filter?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 16]
```diff
@@ -3747,6 +3778,50 @@ public int getCountOfApplications(String filter) throws IdentityApplicationManag
         return count;
     }
 
+    /**
+     * Retrieve application basic information using the sp metadata property key and value.
+     *
+     * @param key Name of the sp metadata property key
+     * @param value Value of the sp metadata property value
+     * @return
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3786]:
```suggestion
     * @return Filtered applications' basic information.
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java (hunk 15 above)
Comment (AnuradhaSK) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:2279]: Seems this is fixed in the same PR. Hence ignore the comment

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**